classpath/java/util/Collections.java

```diff
@@ -500,10 +514,41 @@ public boolean hasNext() {
     }
 
     public void remove() {
-      if (toRemove != -1) {
-        list.remove(toRemove);
-        index = toRemove - 1;
-        toRemove = -1;
+      if (prevIndex != -1) {
+        list.remove(prevIndex);
+        index = prevIndex - 1;
+        prevIndex = -1;
+      } else {
+        throw new IllegalStateException();
+      }
+    }
+
+    @Override
+    public int nextIndex() {
+      if(hasNext()) {
+        return index+1;
+      }
+      return index;
+    }
+
+    @Override
+    public void add(T e) {
+      list.add(index, e);
+    }
+
+    @Override
+    public int previousIndex() {
+      if(hasPrevious()) {
+        
```

Comment:
nope, we should probably be returning index-1, or maybe index depending on how we are keeping track.  Feel free to fix it, or I will once I get a chance.